freqtrade/freqtradebot.py

```diff
@@ -582,8 +593,6 @@ def execute_entry(self, pair: str, stake_amount: float, price: Optional[float] =
             amount = safe_value_fallback(order, 'filled', 'amount')
             enter_limit_filled_price = safe_value_fallback(order, 'average', 'price')
 
-        # Fee is applied twice because we make a LIMIT_BUY and LIMIT_SELL
-        fee = self.exchange.get_fee(symbol=pair, taker_or_maker='maker')
```

Comment:
I think keeping this "inline" will make more sense - as it does not make sense to assign the close_fee a different value - as the value is in ~80% of the cases wrong.

CCXT is returning the exchange default value as far as i know (not the one tied to your particular account including rebates, VIP levels etc.).
Therefore, the moment you use your account "a bit" - your real fees will be different (e.g. Paying in BNB on binance gives you quite some rebates).

As such - there is logic in place to make the close-fee guess to be identical than the entry-fee (we know you paid 0.05% instead of the default 0.1% - so it's pretty safe to assume you'll get the same rebate on closing, too).
So assigning close_fee here will have only temporary impact - with even more questionable benefits (even if we know you're using market orders to exit - the fee will still be different).

Reply:
> as it does not make sense to assign the close_fee a different value

Why? What if it is a different value

> Therefore, the moment you use your account "a bit" - your real fees will be different (e.g. Paying in BNB on binance gives you quite some rebates).

How is anything related to that changed? I don't think I changed anything related to custom fee rates in the config. If it's wrong now, it would have been wrong before too, unless something calls `Trade.update_fee` that I'm not seeing

Reply:
> Why? What if it is a different value

It's overwritten the moment the order fills - so you write it here and it's valid potentially only for seconds.
Also the opening fee will be wrong in the same 80% of the cases if you trade live - simply because you use real money - and exchanges have different fees depending on the usage of the account - which ccxt will not return until you placed an order.

> Why?

I don't think it's wrong now - but it doesn't sense to assign different fees for open and close if different order types are used - if we know for a fact that it'll be overwritten the moment the order fills with the fee the order filled with.

As such - I'd have kept it as "get open fee - assume that same fee for closing" (which would align it to "after the order filled", too). 
It's as good a guess as any - and just as wrong as the new approach.

Reply:
In dry run it's calculating the wrong fees, that's how I noticed it, and that's why I'm changing it, strategies that use market orders don't have accurate dry runs

Reply:
![](https://i.imgur.com/7a1evg2.png) Those are all market orders, so they're going to be maker, and should be 0.004, if my open orders are limit though with exit orders being market (which is common when being slow to get it but quick getting out) then the fee rate for open orders would be 0.002 but 0.004 for exit orders

Reply:
> and exchanges have different fees depending on the usage of the account 

CCXT actually has the fee tiers for a number of exchanges, it doesn't have them for binance spot, but I could fix that, it wouldn't matter unless the user gives their tier anyway because dry run doesn't access private methods

Reply:
it wasn't clear that you're talking dry-run only with this - what I'm talking about above is live.

the updating of fees should not happen in dry-run - so there we'd be safe (my main thinking was about live though - where it'll be overwritten as fees are updated).

Reply:
> it wasn't clear that you're talking dry-run only with this - what I'm talking about above is live.
> 
> the updating of fees should not happen in dry-run - so there we'd be safe (my main thinking was about live though - where it'll be overwritten as fees are updated).

Does it make sense to you that limit orders would mostly be maker, and market orders would mostly be taker? 

Also, I added the order book side logic to the get_fee method

Reply:
even for live they shouldn't be the same, if ones a taker order and the other is maker they'll be different, we need to fetch both for live

Reply:
My point is a different one - and obviously you didn't get that yet.

I'm not arguing that the fees are identical - what i'm saying is that whatever we assume a fee to be **before an order fills** is an estimation - which is most likely wrong (wrong in ~80% of the cases - at least on binance).

---

while the fees may be different - they will ALSO be different by user account.
you may not be paying the same fees on binance than i do, because my volume might be above yours, or because I'm holding more BNB - or because i have fewer referals or whatever else counts into this equation (it's different by exchange).

While you're talking about "tiered" fees - that's all nice and well - but we'll still not get the **correct fees** until an order is executed (fees estimated based on the user's account parameters). 
That's not a ccxt limitation, but an exchange API one - as I'm yet to see such an endpoint from the exchange.


---

So in a **mixed** case (limit for entry, market for exit) 
- we will take an assumption at entry (this is what you're messing with now) (say - `0.001 for entry`, and `0.002 for exit`).
- Once the order fills - we'll know the real rate the exchange applied (this may or may not be different to the value we get via `get_fee()` in reality, we did pay `0.0005` (half of the original fee) - and in BNB (this is not relevant, we convert it to a ratio).

As the order filled - we now know what the entry rate was - so `update_fee` will update the close-rate to `0.0005`.
This means, your prior estimation (`0.002` for exit) was pointless  as it is overwritten.

Now we have 2 possibilities at this point - we either assume that the initial assumption was correct (`0.002` for the exit) - so we remove the "set close fee" line.
or we assume that the adapted fee is more appropriate - so we leave it as is, and assume `0.0005` is our new exit fee.

The problem is - both are potentially wrong - as the ordertype is different, we might encounter `0.0015` as exit-fee (maybe rebate is lower for market orders).

This is why i think setting the close-fee differently is pointless (at least when looking at live) - as we simply do not know the real fee until an order is executed (and even then, it's only a guess for the next order, as circumstances could change and therefore different rates may apply for the exit).

Reply:
> This is why i think setting the close-fee differently is pointless (at least when looking at live) - as we simply do not know the real fee until an order is executed (and even then, it's only a guess for the next order, as circumstances could change and therefore different rates may apply for the exit).

Well a better solution would be to multiply the open fee by maker/taker if open is taker and close is maker, and taker/maker if open is maker and close is taker

> This will depend on the pricing side - limit orders can also pay taker fees if it becomes a "marketable limit order" - which basically means that a buy price above the spread was selected. Then we define the "worst" possible price the order will fill - but if the current best ask is below the price, then we'll get that price (but pay taker fees).

Yeah I know, but the best we can do in dry_run is assume maker or taker based on the config settings in ["bid_strategy"]["price_side"] and ["ask_strategy"]["price_side"]

My fixes in this PR are for dry run though

Reply:
> As the order filled - we now know what the entry rate was - so update_fee will update the close-rate to 0.0005.
This means, your prior estimation (0.002 for exit) was pointless as it is overwritten.

I'm not sure if I know exactly what code your talking about, but in Trade.update_fee, I made it so that the exit fee gets set to the open fee, but the open and exit fee are proportial as there were before, so setting the close_fee with the `0.0002` value because it helps to determine what the size of the close fee is compared to the open fee